### ISSUE TITLE
Pictures cannot be uploaded

### ISSUE THREAD

**Issue by mcsSolutions** [HomeSeer/Plugin-SDK#107]
Version 4.1.1.0

Postback data on large forms such as is used to upload a megabyte picture is not posted to the plugin.  Two examples with the postback from wireshark are attached the "pictureentry" one is a 1.6 MB post.  The "others" one is a 4KB post.  The smaller one is made available to the plugin's postback function.  The larger one goes into the bit bucket.

These pages are produced when the form uses "enctype=multipart/form-data and input type=file".  The plugin is able to decode the raw data, but obviously not able to decode it if it is never made available.

![Multipart](https://user-images.githubusercontent.com/1480699/88106334-2d8e0a00-cb5a-11ea-9c29-4ebe9dd2d254.PNG)
![Other](https://user-images.githubusercontent.com/1480699/88106464-60380280-cb5a-11ea-8ffc-738630d950c5.PNG)

If other HS4 Server methods are available to handle picture/file uploads then an alternate mechanism can be implemented in the plugin if information is available on doing it.

**Comment by mcsSolutions** [comment]
Additional testing shows that only the first frame of 65536 bytes is returned when approx 66000 bytes were posted.  This means sometimes partial data is forwarded to plugin and other times no data is forwarded for the cases where more than 65536 are posted.

**Comment by rjhdakota** [comment]
Do you have a way for me to test this? I used postman and posted 2248808 bytes and they all came through to the plugin. So it might have something to do with the way the file is uploaded.

**Comment by rjhdakota** [comment]
All I really need is the HTML page you are using, I can add that to my test plugin.

**Comment by mcsSolutions** [comment]
The attached zip contains an image of the page and the HTML that produced it.  The essential element is any of the input=file form tags and selecting a file that is more than 65K such a 1 MB jpg.

**Comment by rjhdakota** [comment]
No zip came through, maybe you can add it from github.com rather than email.

**Comment by mcsSolutions** [comment]
Attaching zip here
[Picture.zip](https://github.com/HomeSeer/Plugin-SDK/files/4997195/Picture.zip)

**Comment by rjhdakota** [comment]
That page you included is incomplete and includes all the content twice. It looks like an HS3 page. Is this issue with an HS3 plugin? Is there a plugin I can load to re-create the issue? I need some exact steps so I can re-create it here.

**Comment by mcsSolutions** [comment]
[Sample-Plugin-VB-FileInput.zip](https://github.com/HomeSeer/Plugin-SDK/files/5009862/Sample-Plugin-VB-FileInput.zip)



Attached is the HS4 sample plugin with sample-blank.html edited to add the file input tag.  The plugin code was only changed in the Postback function to show the length of the postback.

Two  form tags are in sample-blank.html.  The first is a button that submits as expected.  The second (File Input), when clicked and file selected has no postback to the plugin.  The same html construct used in HS3 mcsSprinklers HTML does generate a postback where the plugin decodes the multipart form to save the uploaded image.

While not include in this test code, it is also possible to produce a form that has a large (> 65K) postback that does not use the File tag.  Only the first block of the postback is delivered so the input provided back to the plugin has been truncated.

**Comment by rjhdakota** [comment]
In regards to what you posted, I have loaded it. The file upload requires some javascript to work, it will not work with straight HTML. In HS3 we have a jqfileupload control that you probably used on your form. You could use that on this form also. Take a look at the HTML generated by HS3. You can also look at the HS4 HTML file html/includes/actions.html line 368. I use the HS3 jqfileupload there to upload audio files.

As for the >65K issue, if you are uploading a file then I would just use the jqfileupoload as I know that handles any size file.

**Comment by mcsSolutions** [comment]
mcsSprinklers for HS, HS2 and HS3 has always used raw HTML for its pages.  HS4 is the first version that has not supported the ability to use the “file” tag and large multipart forms.  In the case of the large text postback the postback is a standard browser page with various input controls,  It uses multipart form because one of the potential inputs is a picture upload.  It is not something that could be represented as a file upload.

 

I understand that multipart forms may not be on your roadmap for HS4.  I will continue with mcsSprinklers conversion with recognition that there will be features that cannot be supported in HS4.

**Comment by rjhdakota** [comment]
The server supports multi-part forms. The webserver has not changed, but the data routing has changed due to the new pages. So in the sample you sent, you expect the second button to post the image that you select? I can try the page with HS3, I assume it posts as expected there?

**Comment by mcsSolutions** [comment]
I attempted to provide some data for the incomplete postback data and determined that the issue occurs on both HS3 and HS4.  I put a tag at the end of the form named "complete" so the plugin is able to assess if the complete vs. partial form was returned.  In the console of the postback function (HS3 PagePut) I displayed the length and last 100 characters.  I could repeat both the good and fail cases and the lengths did not change in both cases (4380 bytes and 4911 bytes), but I could not correlate when a shortened postback with anything.  It seems random.  The Console output for the two cases is shown below.  I doubt if it is worth investigating further on your side since the issue exists in HS3 too and I have a means to detect the failure.

I will also investigate using javascript per actions.html.  I did capture the HTML for a play audio event.  It is not clear to me what information is actually returned to the plugin as a result of doing the upload file click.  Does the HS server store the file somewhere?  Is a postback done when the upload is complete?  Are MIME types other than audio/* supported? 


FAIL/SHORT CASE
=================
Postback length 4380 ---WebKitFormBoundaryOAnR5SZQ2wyWqGyb
Content-Disposition: form-data; name="SQLPassword"


-----

GOOD/NORMAL CASE
=================

Postback length 4911
Content-Disposition: form-data; name="complete"

1
------WebKitFormBoundaryeHcIobmfwefeEaAn--

**Comment by rjhelmke** [comment]
Is this still an issue for you, looks like maybe it's not resolved?

**Comment by mcsSolutions** [comment]
There were too many obstacles for converting mcsSprinklers plugin to HS4.  Any GitHub issues reported about mcsSprinklers conversion can be closed.

**Comment by rjhelmke** [comment]
Ok, understood. I would very much appreciate it if you could email me about the issues that prevented you from converting your plugin. We are constantly updating the software so this would be good information to have. -Rich

**Comment by mcsSolutions** [comment]
The initial problem was the lack of Device Code.  mcsSprinklers was developed with HS1 and Rain8-X10 as the launch interface.  Without the ability to get the DC from HS4 the identification of all devices already setup in HS3 was not possible.

 

DeviceType (String) was also not available.  This was used to distinguish Valves from Status devices.

 

mcsSprinklers has over 100 status devices.  This many features in a single HS4 tile would not provide a good user experience even if HS4 supported it.

 

I had design approaches to overcome the above, but where I gave up in July 2020 was with the HTTP server.  mcsSprinklers was designed with its own web server so using it as a plugin means it needed to fit with the HS server capability.  Pictures were uploaded and then post-processed.  I was able to get this to work with HS3, but not HS4.  I believe this is the GitHub issue in this thread.

 

If I recall correctly HS combined querystring and post data while mcsSprinklers server kept the two separate as there was overlap in the keys.

 

There was extensive use of hyperlinks to navigate within mcsSprinklers pages.  Hyperlinks existed on icons and multiple icons were used on a single status device.   Some of the hyperlinked URLs were dynamically generated.  This made is difficult to register pages with HS server.  This part should have been able to recoded, but never got to the point of doing it.  At the time of starting with HS4 the entire JUI thing was foreign to me.

 

There were more than 20 pages that were developed in HTML and most were dense to pack much information into a small space.  Putting the bootstrap wrapper around them killed their effectiveness and in some cases the control disappeared due to table cell size limits vs. bootstrap expectations.

 

Overall it got to the point were there was just too much.  I recognize that some of the previously depreciated capability has been restored, but even now that I am more comfortable with HS4 API, I recognize how much work it would be.
